kuksa-sdk/src/main/kotlin/org/eclipse/kuksa/extension/VssSpecificationCopyExtension.kt

```diff
@@ -0,0 +1,131 @@
+import org.eclipse.kuksa.proto.v1.Types.Datapoint.ValueCase.*
+import org.eclipse.kuksa.vsscore.model.VssProperty
+import org.eclipse.kuksa.vsscore.model.VssSpecification
+import org.eclipse.kuksa.vsscore.model.findHeritageLine
+import org.eclipse.kuksa.vsscore.model.heritage
+import org.eclipse.kuksa.vsscore.model.variableName
+
+/**
+ * Creates a copy of the [VssSpecification] where the whole [VssSpecification.findHeritageLine] is replaced
+ * with modified heirs.
+ *
+ * Example: VssVehicle->VssCabin->VssWindowChildLockEngaged
+ * A deep copy is necessary for a nested history tree with at least two generations. The VssWindowChildLockEngaged
+ * is replaced inside VssCabin where this again is replaced inside VssVehicle.
+ *
+ * @param changedHeritageLine the line of heirs
+ * @param generation the generation to start copying with starting from the [VssSpecification] to [deepCopy]
+ * @return a copy where every heir in the given [changedHeritageLine] is replaced with a another copy
+ */
+fun <T : VssSpecification> T.deepCopy(changedHeritageLine: List<VssSpecification>, generation: Int = 0): T {
+    if (generation == changedHeritageLine.size) { // Reached the end, use the changed VssProperty
+        return this
+    }
+
+    val childSpecification = changedHeritageLine[generation]
+    val childCopy = childSpecification.deepCopy(changedHeritageLine, generation + 1)
+    val parameterNameToChild = mapOf(childSpecification.variableName to childCopy)
+
+    return copy(parameterNameToChild)
+}
+
+/**
+ * Creates a copy of a [VssProperty] where the [VssProperty.value] is changed to the given [Datapoint].
+ *
+ * @param datapoint the [Datapoint.value_] is converted to the correct datatype depending on the [VssProperty.value]
+ * @return a copy of the [VssProperty] with the updated [VssProperty.value]
+ *
+ * @throws [NoSuchElementException] if the class has no "copy" method
+ * @throws [IllegalArgumentException] if the copied types do not match
+ */
+fun <T : Any> VssProperty<T>.copy(datapoint: Datapoint): VssProperty<T> {
+    with(datapoint) {
+        val value: Any = when (valueCase) {
+            STRING -> string
+            BOOL -> bool
+            INT32 -> int32
+            INT64 -> int64
+            UINT32 -> uint32
+            UINT64 -> uint64
+            FLOAT -> float
+            DOUBLE -> double
+            STRING_ARRAY -> stringArray.valuesList
+            BOOL_ARRAY -> boolArray.valuesList.toBooleanArray()
+            INT32_ARRAY -> int32Array.valuesList.toIntArray()
+            INT64_ARRAY -> int64Array.valuesList.toLongArray()
+            UINT32_ARRAY -> uint32Array.valuesList.toIntArray()
+            UINT64_ARRAY -> uint64Array.valuesList.toLongArray()
+            FLOAT_ARRAY -> floatArray.valuesList.toFloatArray()
+            DOUBLE_ARRAY -> doubleArray.valuesList.toDoubleArray()
+
+            VALUE_NOT_SET -> 0
+            else -> throw NoSuchFieldException("Could not convert value: $value to actual type")
```

Comment:
this is the perfect example of where an exhaustive-when could make sense :) 
however due to the "else"-branch we lose the advantage.

Reply:
Yeah it is not a sealed enum because of java. But still fine, if a new value is introduced we will find out pretty quick because of the exception.^^

Reply:
If you remove the else and add a branch for null (and throw an exception there) it is exhaustive, right?

Reply:
Does not work for the fixed version with the nested when case.